lisp/tree-sitter-query.el

```diff
@@ -0,0 +1,119 @@
+    (overlay-put overlay-added 'evaporate t)
+    ;; set the match-face as the face of the overlay
+    (overlay-put overlay-added 'face match-face)
+    ;; put the name of the capture in the help-echo, if any
+    (unless (string= capture-name "")
+      (overlay-put overlay-added 'help-echo capture-name))))
+
+(defun tree-sitter-query--eval-query (patterns)
+  "Evaluate a query PATTERNS against the target buffer."
+  (with-current-buffer tree-sitter-query--target-buffer
+    ;; clean the target buffer of overlays
+    (remove-overlays)
+    (let* ((query (ts-make-query tree-sitter-language patterns))
+           (root-node (ts-root-node tree-sitter-tree))
+           (matches (ts-query-captures query root-node nil nil))
+           (nextface)
+           (capture-name))
+      (if (> (length matches) 0)
+          (progn
+            ;; reset counter
+            (setq tree-sitter-query--match-highlight-number -1)
+            ;; iterate all matches and highlight them with an underline
+            (cl-loop
+             for submatches across matches
+             do (cl-loop
+                 for match on submatches
+                 do (unless (string= capture-name (car match))
+                      (setq capture-name (car match))
+                      (setq nextface (tree-sitter-query--get-next-match-highlight-color)))
+                 (tree-sitter-query--highlight-node match nextface))))
+        (message "[ERR] no matches found or invalid query")))))
+
+(defun tree-sitter-query--after-change (&rest args)
+  "Run evaluation of pattern in current buffer for every change made by the user, ignoring ARGS."
+  (with-current-buffer (get-buffer "*tree-sitter-query-builder*")
+    (let ((pattern (buffer-substring-no-properties (point-min) (point-max))))
+      (with-demoted-errors
+          (tree-sitter-query--eval-query pattern)))))
+
+(defun tree-sitter-query--clean-target-buffer ()
+  "Remove all overlays if the builder buffer happens to be killed"
+  (with-current-buffer tree-sitter-query--target-buffer
+    (remove-overlays)))
+
+(defun tree-sitter-query-builder ()
+  "Provide means for developers to write and test tree-sitter queries.
+
+The buffer on focus when the command is called is set as the target buffer"
+  (interactive)
+  (let* ((target-buffer (current-buffer))
+         (builder-buffer (get-buffer-create "*tree-sitter-query-builder*"))
+         (builder-window-is-visible (get-buffer-window builder-buffer))
+         (builder-window))
+    (when (eq target-buffer builder-buffer)
+      (error "This buffer cannot be use as target buffer"))
+    (unless builder-window-is-visible
+      (setf builder-window (split-window-vertically -7))
+      (with-selected-window builder-window
+        (switch-to-buffer builder-buffer)))
+    (with-current-buffer target-buffer
+      (unless tree-sitter-mode
+        (tree-sitter-mode))
+      (add-hook 'after-change-functions 'tree-sitter-query--after-change nil t)
```

Comment:
IMO when you have boolean args like this it's better to use keywords to indicate what they mean. `:local` is better readable than plain `t`.